Adding racs2 /cfs-ros2 bridge demo by JAXA on spaceros

yuyuqq opened the pull request:
**Title**: Adding racs2 /cfs-ros2 bridge demo by JAXA on spaceros 

**Description**:
This pull request adds the racs2_demos_on_spaceros demo to the SpaceROS project. (racs2 /cfs-ros2 bridge demo by jaxa)
(Issue #26 )

**Changes include**:
- Added new demo module racs2_demos_on_spaceros.

**Purpose**:
This addition aims to demonstrate a function and usage of  RACS2 bridge in the SpaceROS project, providing a tool for system design and showcasing functionalities.

**Testing**:
All new features were tested in the provided simulation environment to ensure proper functionality and integration. No issues were found.

**Impact**:
These changes are backward-compatible and do not affect existing functionalities.

**Additional Information**:
For more details, please refer to the included README file or the project documentation.

ivanperez-keera commented:
I just added myself as reviewer for this PR. I'd like to work with @yuyuqq on this.

dsobek reviewed (commented):
I started reviewing however I got blocked with confusion: Is having the `docker` directory necessary here? It looks like a copy of the demos from the [docker repo](https://github.com/space-ros/docker/tree/main). 

To reduce duplicate code, my first thought is that the docker repo should be a submodule, however I noticed that space_robots clones this repository via vcs within it's Docker workflow, so this is looking circular.

My second thought is that the racs2 demo docker image should be built from the docker repository and any lower level packages/nodes/cfs applications should be here in the demo repository, however this is conflicting with the [demo README](https://github.com/space-ros/demos/blob/main/README.md).

Overall I could use some clarity on the purposes of the docker and demo repositories @ivanperez-keera.

ivanperez-keera commented on racs2_demos_on_spaceros/Dockerfile:
```diff
@@ -0,0 +1,125 @@
+RUN python3 -m pip install protobuf==3.20.0 websockets==12.0
+
+# Get the cFS source code
+RUN git clone --recursive -b v6.7.0a https://github.com/nasa/cFS/ cfs
```
I think this kind of clone is fine in this case because we are cloning a specific tag. The code won't change unless the tag changes, and that shouldn't happen because it's a release tag.

If it was a rolling HEAD, then I agree that we might want to do it differently.

yuyuqq replied:
As Ivan suggested I did not change on this.

ivanperez-keera commented:
We met today with the JAXA team to discuss this pull request and walk them through some of the changes. They'll be working on that and submitting an updated version to this same PR.

yuyuqq commented on racs2_demos_on_spaceros/Dockerfile:
I have modified the pull request as suggested to reduce the number of clone operations in the Dockerfile, making it simpler. This should align better with typical Dockerfile practices while still providing step-by-step documentation for the RACS2 bridge.

yuyuqq commented on racs2_demos_on_spaceros/Dockerfile:
```diff
@@ -0,0 +1,125 @@
+
+# Get the RACS2 source code
+WORKDIR ${RACS2_DEMO_DIR}
+RUN git clone https://github.com/jaxa/racs2_bridge
```
Added a tag called "v1.1" and made it clone from v1.1.

yuyuqq replied:
Changed a tag to "v1.1.1" which is a bug fix version of v1.1.
See https://github.com/jaxa/racs2_bridge/issues/22

yuyuqq commented on racs2_demos_on_spaceros/README.md:
````diff
@@ -0,0 +1,36 @@
+```
+You will need three terminals to run demo.
+
+For instractions after entering docker, please read docker/racs2_demos_on_spaceros/README.md
````
Merged README.md, thus this description is removed.

yuyuqq commented:
The PR is updated.

> I started reviewing however I got blocked with confusion: Is having the docker directory necessary here?

We removed the docker directory, and copied necessary files in the docker file. This should have solved the "looking circular" problem. 

> My second thought is that the racs2 demo docker image should be built from the docker repository

Agreed, we made docker-build from the docker repository with necessary changes.

Also logs are sorted out as Ivan suggested.

ivanperez-keera commented:
Thanks @yuyuqq Will take a look again.

dsobek commented on racs2_demos_on_spaceros/README.md:
````diff
@@ -0,0 +1,93 @@
+Launch the rover demo (calling Terminal 1):
+```bash
+source install/setup.bash
+ros2 launch mars_rover mars_rover.launch.py
````
FYI I'm on apple silicone using a VM and I had to set an environment variable to enable software rendering in order to get his working.
```
export LIBGL_ALWAYS_SOFTWARE=1
```

@ivanperez-keera Do we have a good place to document this. As far as I'm aware, this affects any gazebo simulation for setups like mine.
https://github.com/gazebosim/gz-sim/issues/1841#issuecomment-1404172345

yuyuqq replied:
@dsobek Thanks for the clarification.
I assume no changes are required on my side regarding this point.

I am currently working on a few code updates based on other comments and will push them shortly.

yuyuqq commented:
### Updated RACS2 bridge to v1.1.1 to incorporate bug fixes.

This is the main purpose of this PR. 

- Integrated the latest bridge version (v1.1.1 instead of v1.1)

To ensure compatibility with the updated bridge, several adjustments were made to the demo environment:

- Updated Dockerfile to use new curiosity_rover-based structure (minimal changes to reflect updated layout)
- Cleaned up old directories (config/, worlds/) that are no longer in use
- Integrated `curiosity_gazebo.launch.py` into the main launch file
- Adjusted CMakeLists.txt and package.xml accordingly
- Updated README.md to reflect new build and run procedures

The core logic and node implementations remain mostly unchanged from the previous version. Only structural updates necessary to support the new bridge version were applied.

Thanks for the feedback so far — remaining small cleanup items will be addressed next.

yuyuqq commented on racs2_demos_on_spaceros/cFS/apps/run_app/fsw/src/run_app.c:
```diff
@@ -0,0 +1,436 @@
+void set_non_canonical_mode(int fd) {
+    struct termios termios;
+
+    // 現在の端末設定を取得
```
I made the comments in English.

yuyuqq commented on racs2_demos_on_spaceros/cFS/apps/run_app/fsw/src/run_app.c:
```diff
@@ -0,0 +1,436 @@
+
+    SAMPLE_TAKLKER_Init();
+
+    struct termios original_termios;
```
I confirmed it is not used within the function, and removed it.

yuyuqq commented on racs2_demos_on_spaceros/cFS/apps/run_app/fsw/src/run_app.c:
```diff
@@ -0,0 +1,436 @@
+        // message.linear->z = 0.0;
+        // message.angular->x = 0.0;
+        // message.angular->y = 0.0;
+        // message.angular->z = 0.0;
```
I removed them.

yuyuqq commented on racs2_demos_on_spaceros/README.md:
````diff
@@ -0,0 +1,93 @@
+./run.sh
+```
+
+Depending on the host computer, you might need to remove the ```--gpus all``` flag in ```run.sh```, which uses your GPUs.
````
I changed from "remove" to "remove/add".

yuyuqq commented:
The remaining small cleanup items are fixed.

ivanperez-keera commented:
Thank you, @yuyuqq . I'll take a look.

yuyuqq commented:
The GitHub Actions check is currently failing, but I believe that the error is outside the scope of this PR.

My changes are limited to `racs2_demos_on_spaceros`, while the failure appears to occur in `demo/curiosity_rover/build.sh` due to missing or misnamed apt packages unrelated to this demo.

Let me know if you'd like me to look into that separately.
